### DISCUSSION TITLE
Why ReadonlyXyz?

### DISCUSSION THREAD

**Discussion by edalex-ian** [gcanti/fp-ts#1609] [Q&A] [answered]
In responding to #1608 @mlegenhausen highlighted to me the abundance and use of the ReadonlyXyz type classes (`ReadonlyNonEmptyArray`, `ReadonlyRecord`, `ReadonlyMap`, etc.). But I was under the impression that the non-ReadonlyXyz type classes were following the immutable paradigm. Is that not correct?

Or, if that is correct then what's the intention behind these? Documentation purposes?

**Comment by mlegenhausen** [comment, marked as answer]
Me again 😅

> But I was under the impression that the non-ReadonlyXyz type classes were following the immutable paradigm

This only holds if you use `fp-ts` functions for data manipulations, but not for the functions that are defined on the non-readonly interfaces. An `Array` still has a `push` function which will mutate the array. On an `ReadonlyArray` the `push` function is not defined at all. So using the `Readonly` variants gives you simply more type safety, but with a "cost" that you need to explicitly convert to a mutable array when e. g. calling APIs that expects a mutable data structure.

In `fp-ts@3` the mutable versions of `Array`, `Record`, `NonEmptyArray` and `Map` will be deprecated and will only contain functions to convert from and to readonly versions of these data structures. See Issue https://github.com/gcanti/fp-ts/issues/1226.

**Reply by edalex-ian** [reply]
Thanks again @mlegenhausen .

Just to make sure I understand you, where you say:

> An `Array` still has a push function function which will mutate the array.

Here you're referring to standard [JS Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array), and so really [fp-ts Array](https://gcanti.github.io/fp-ts/modules/Array.ts.html) is not changing the underlying type - just providing a thin veneer over it to allow FP idioms to be used. Whereas with `ReadonlyArray` we now have a different type, and so fp-ts has full control?

**Reply by mlegenhausen** [reply]
> Here you're referring to standard JS Array, and so really fp-ts Array is not changing the underlying type - just providing a thin veneer over it to allow FP idioms to be used.

Yes, and if you take a look at the `Array` source you will not find any `Array` definition cause the standard typescript Array datatype is used.

> Whereas with ReadonlyArray we now have a different type, and so fp-ts has full control?

No, it is just the same as with the `Array`. `ReadonlyArray` is a datatype shipped with TypeScript, but it simply fits better with functional programming due to the guarantees you get from immutability.

**Reply by edalex-ian** [reply]
Ah! I'd missed that typescript had a `ReadonlyArray` - and I now see [since 3.4 you can also just use the `readonly` modifier](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#improvements-for-readonlyarray-and-readonly-tuples).

Thanks for the clarification @mlegenhausen :+1:
